README.md
# HAVP - HTTP Antivirus Proxy

![HAVP-Logo](http://www.havp.org/wp-content/uploads/2020/10/HAVP.png)

## Short Description
HAVP (HTTP Antivirus Proxy) is a HTTP proxy with an antivirus scanner. It supports the free ClamAV , but also commercial solutions e.g. Kaspersky, Sophos and F-Prot. The main aims are continuous, non-blocking downloads and smooth scanning of HTTP traffic. Havp antivirus proxy has a parent and transparent proxy mode. It can be used with squid or standalone

Further information can be found at
<http://havp.org>

## UPGRADING

Just install HAVP normally. Your config will be preserved, but check
havp.config for possible new options. Templates are overwritten, so if
you have your own, make sure it is not in any default directory.


## BASIC INSTALLATION

HAVP has been tested only with GCC.
Other compilers like Sun Studio have some problems currently.

Installation:

```
   ./configure    (if you don't want /usr/local, use --prefix=/other/path)
   make
   make install
```

You can use the following path options in configure:

```
  --prefix         base directory, default "/usr/local"
  --sbindir        location of havp-binary, default "$prefix/sbin"
  --sysconfdir     location of etc, default "$prefix/etc" (+ /havp)
  --localstatedir  location of pidfile, default "/var" (+ /run/havp)
```

  Also `make install DESTDIR=/tmp/havp` is supported for helping
  in creating packages etc.

It is recommended to create a havp user:

```
  # groupadd havp 
  # useradd -g havp havp
```

Check the configfile: `/usr/local/etc/havp/havp.config`

If Linux is used, you need to enable mandatory locking for the partition
where your tempfiles are located. Solaris supports mandatory locking
without these extra steps:

  The default location for logfiles is `/var/spool/havp`
  Don't use mandatory locking for `/`

  Using tmpfs might have some problems, make sure you test it properly.
  Add mand-option to `/etc/fstab` so it will stay after reboot e.g:
  
  `echo "none /var/spool/havp tmpfs mand,nodev,nosuid,noexec,nodiratime,size=50M 0 0" >> /etc/fstab`

  NOTE: Mandatory locking could make it possible for evil local accounts
  to hang the system. You should run HAVP anyway on non-public server.

Make sure the directories you are using have correct permissions:

```
  # chown havp /var/spool/havp /var/log/havp /var/run/havp
  # chmod 700 /var/spool/havp /var/log/havp /var/run/havp
```
Start havp:
```
  # /usr/local/sbin/havp -c /path/to/config
```
You can also install rc-script to your system from sources etc/init.d.

If you have problems check the logfiles:
```
  /var/log/havp/havp.log
  /var/log/havp/access.log
```
More information and help can be found at HAVP forum: http://havp.hege.li/



## OS SPECIFIC INSTRUCTIONS

Linux:
------

Use GCC 3.4+.

Solaris 9:
----------

You may need lots of swap space if you use library scanners (ClamAV and
Trophie). It wants to reserve it even when it is not really used. If there
is not enough, you will get fork errors. Worst case formula: (20MB *
USEDLIBRARYSCANNERS) * (USEDSCANNERS + 1) * SERVERNUMBER.

GCC 3.4.2 from sunfreeware.com is recommended.

You may need to fix GCC headers like this:
```
  # cd /usr/local/libexec/gcc/*/3.4.2/install-tools
  # ./mkheaders
```
Solaris 10:
-----------

Swap space is not an issue anymore.

Use GCC 3.4.x that comes bundled at `usr/sfw/bin/gcc`.
It is installed from SUNWgcc package.

FreeBSD:
--------

Use GCC 3.4+ from ports. FreeBSD does not support mandatory locking, which
means KEEPBACK settings can not be used (only TRICKLING is supported). This
means everything is first downloaded fully and only then sent to client.

You need to use `--disable-locking` option to compile.



## SCANNER SPECIFIC INSTRUCTIONS

ClamAV 
------

Library is used directly, so there is no need for clamd running.

If you choose to use clamd (which is not recommended as library support has
less overhead), you need to enable AllowSupplementaryGroups in clamd.conf,
and add clamav user to havp group.



####== NOTICE: ==
You must check your antivirus license before using HAVP with commercial
scanners. Usage might not be allowed. We do not give any warranty!


Kaspersky 
---------

Tested with aveserver daemon found in Linux File Server and Linux Mail
Server package.

You should set ReportLevel=1 at [aveserver.report] section, so log will not
fill disk.


Trend Micro (Trophie)
---------------------

/etc/iscan must point to the directory where libvsapi.so and
virus patterns are located. Create link if necessary.

Trend library is used directly, so daemon is not required to be running.
You should naturally run some pattern update script, if Trend itself is
not running.


AVG 
---

Recommended changes to avg.conf (version 7.5): 
```
 [AvgCommon] 

 heuristicAnalysis = 1 
 processesArchives = 1 

 [AvgDaemon] 

 # Raise number of daemons atleast equal to SERVERNUMBER/MAXSERVERS
 numOfDaemons = xx 
```

F-Prot 
------

Supported.


NOD32
-----

Tested with Linux Mail Server and Linux File Server packages.
File Server version can not display virus names.

For version 3.0+, see settings in /etc/esets/esets.cfg (num_thrd etc). Also
you want to disable syslogging.


Sophos (Sophie)
---------------

You need to make sure Sophie is working first, you can get it from:
http://www.clanfield.info/sophie/

Change user or group to havp user in sophie.cfg, so it can read tempfiles.
Also change maxproc value to atleast SERVERNUMBER/MAXSERVERS value!


Avast!
------

Linux/Unix Servers version is required.

Recommended changes to avastd.conf:
```
 # Raise number to atleast equal of SERVERNUMBER
 daemoncount = XX
 # Raise number to atleast equal of MAXSERVERS
 maxdaemoncount = XX
 archivetype = A
 testall = 1
 testfull = 0
```

Arcavir
-------

Start arcavird with enough processes, like "arcavird 16".


DrWeb
-----

Recommended changes to drweb32.ini: 
```
; Raise number to atleast equal of SERVERNUMBER 
MaxChildren = xx 
PreFork = Yes
```



ChangeLog
*HAVP 0.93 released (2019/02/02)
- Add support for clamav 0.101: Sebastian Andrzej Siewior <sebastian@breakpoint.cc>
- Enable LFS & fix autoreconf: Sebastian Andrzej Siewior
- Add SSLTIMEOUT config option: Rene Mayrhofer, adapted from a patch by Heiko Schlittermann
- Fix missing re-opens of logfiles after signal(7): Sebastian Andrzej Siewior
- Remove the VIA header completely if disabled: Sebastian Andrzej Siewior
- Change default PID file and SCANTEMPFILE location: <mfladischer@home.fladi.at> adapted and cleaned up by Rene Mayrhofer
- Change default clamd socket: Scott Kitterman
- Update transparent proxy documentation link: Sebastian Andrzej Siewior

*HAVP 0.92a - Security Fix
 - Add missing dot to sourceforge for safety

*HAVP 0.92 released
 - Add SCANMIME and SKIPMIME options
 - Add TIMEFORMAT option
 - Add VIRUSLOG option
 - Add PARENTUSER/PARENTPASSWORD (thanks to James Brotchie)
 - DISABLELOCKINGFOR default has changed in favor of ClamAV 0.96,
   it only contains AVG:ALL now

*HAVP 0.91 released
 - Fix possible segfault on dns lookups (thanks Gavin McCullagh)
 - Fix compiling with gcc 4.4
 - Support AVG version 8.5 (default AVGPORT 54322) (thanks Markus Wigge)
   (note: you may need DISABLELOCKINGFOR AVG:ALL, which is now default)

*HAVP 0.90 released
 - ClamAV library 0.95 support (recompile needed)
 - Support NOD32 version 3 (set NOD32VERSION 30 in config)
 - Add PRELOADZIPHEADER config (Squid 3.x might not work if enabled)
 - Add SYSLOGVIRUSLEVEL config

*HAVP 0.89 released
 - Fix possible retry loop and hang (thanks to Peter Warasin @ endian.it)
 - Always send Via: header, fixes some IIS problems (e.g. MSNBC)

*HAVP 0.88 released
 - ClamAV library 0.93 support (recompile needed, new option CLAMMAXSCANSIZE)
 - CLAMMAXFILESIZE default is now 100MB (so 0.93 even starts scanning big files)
 - Fix random seed issue (ClamAV generated some temporary file errors)
 - Added DISABLELOCKINGFOR config (fix for ZIP handling in ClamAV 0.93)
 - Arcavir version 2008 support (set ARCAVIRVERSION)
 - Log scanner errors to errorlog
 - Relaxed SSL/CONNECT port limits
   (It is _not_ recommended to use --enable-ssl-tunnel, you should use Squid)

*HAVP 0.87 released
 - DrWeb scanner support
 - F-Prot support for v6.0 added (also check FPROTOPTIONS)
 - If false, X_FORWARDED_FOR drops also Via: header for privacy
 - Fix Avast and AVG bugs
 - Templates support <!--url--> and <!--clientip--> tags
 - Uses supplementary groups for user if defined
 - Added TRICKLINGBYTES config
 - Reduced *MAXFILES settings to 50 for performance
 - Add missing HTTP methods (MKACTIVITY, CHECKOUT, MERGE) 

*HAVP 0.86 released
 - Experimental support for chunked Transfer-Encoding, fixes some broken sites
 - Added IGNOREVIRUS config for whitelisting virus names
 - Added CLAMBLOCKBROKEN config
 - Improve detection with AVG, only Viruses were found, not Trojans etc
 - HAVP is killed if database reloading fails for Library Scanner
 - Log URL when crashed scanner process detected, for troubleshooting
 - Build system updated (--prefix --sbindir --sysconfdir --localstatedir)

*HAVP 0.85 released
 - Added support for ClamAV 0.90 library
   (once compiled for it, older library versions not supported)

*HAVP 0.84 released
 - Fix bug in tcp buffer, requests could leak to other clients sometimes
 - Support for Arcavir scanner
 - Support for multiple IPs in hostnames, all are tried if necessary
 - Pass Proxy-Authorization header to parent proxy (thanks Mateus)
 - Ignore scanner errors if MAXSCANSIZE reached (thanks Vittorio)
 - Build system polished a bit (--prefix supported)
 - Added support for old NOD32 scanner (set NOD32VERSION in config)
 - Default for MAXSCANSIZE 5000000, not suggested to be 0 anymore

*HAVP 0.83 released
 - Allow Accept-Encoding header so HTTP compression works
 - Workaround for some ClamAV mmap() troubles (affected BinHex and PDF)

*HAVP 0.82 released
 - Improved ZIP handling (header pre-fetch, scans ZIPs larger than MAXSCANSIZE)
 - Added SCANIMAGES config
 - Ignore RAR errors from clamd
 - Fixes to HTTP header handling
 - Added syslog logging option
 - Signal HUP re-opens logfiles, making rotation possible

*HAVP 0.81 released
 - Fixed binding to low ports (<1024)
 - Fixed FreeBSD, downloads that took longer than TRICKLING did not work

*HAVP 0.80 released
 - Experimental FreeBSD support (no mandatory locking, KEEPBACK not supported!)
 - Avast! scanner support
 - Added MAXDOWNLOADSIZE config
 - Added X_FORWARD_FOR config to control the header
 - Added some archive scanning parameters for Trophie
 - Added TCP support for clamd
 - Ignore RAR errors from ClamAV (use ClamAV-devel if you want to scan RARv3)
 - Fixed bug in socket buffer, sometimes caused nasty effects with POST etc.
 - Fixed KeepAlive for HTTP/1.1 clients, now on by default
 - Access logging format changed a bit

*HAVP 0.79 released
 - MULTIPLE SCANNER SUPPORT! (see havp.config how to enable scanners)
 - NOD32, Sophos and Clamd scanner support
 - Parentproxy error on SSL tunneling is passed to browser
 - Whitelisted sites can use HTTP Range requests (for Windowsupdate..)
 - Added STREAMUSERAGENT/STREAMSCANSIZE config to reduce stream scanning
 - Added SCANNERTIMEOUT option to catch scanners gone wild
 - Added scanning options for ClamLib
 - Some HTTP header handling fixes
 - Init-script updated so it works on Solaris

*HAVP 0.78 released
 - Solaris is now supported too
 - AVG scanner support
 - Kaspersky aveserver socket support (faster)
 - Added TEMPDIR, LOGLEVEL, FAILSCANERROR and WHITELISTFIRST config
 - KEEPBACKTIME config added to complement KEEPBACKBUFFER setting
 - Basic HTTP Keep-Alive support, improves network performance
 - HTTPS/SSL tunneling support (requires configure --enable-ssl-tunnel)
 - FTP is supported and scanned when FTP supporting parent proxy is used
 - Default user and group changed to havp
 - Logging improved
 - Many important and not so important fixes/optimizations

*HAVP 0.77 released

havp/default.h
/***************************************************************************
                          default.h  -  description
                             -------------------
    begin                : 2005/02/12
    last                 : 2019/02/02
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/


#ifndef DEFAULT_H
#define DEFAULT_H

#define VERSION "0.93"

//##############################################################
//Define if you want to rewrite a URL
//#define REWRITE URLRewrite["havp"]="www.havp.org"; URLRewrite["www.havp"]="www.havp.org";

//##############################################################
//Parameters in Configurationfile

#define CONFIGPARAMS \
 "WHITELISTFIRST","TEMPDIR","RANGE", "PRELOADZIPHEADER", "USER","GROUP", \
 "SERVERNUMBER","PORT","BIND_ADDRESS","SOURCE_ADDRESS","KEEPBACKBUFFER", \
 "KEEPBACKTIME","TRICKLING","TRICKLINGBYTES","MAXSCANSIZE","WHITELIST","BLACKLIST","PIDFILE", \
 "DAEMON","TRANSPARENT","LOG_OKS","ACCESSLOG","VIRUSLOG","ERRORLOG","TIMEFORMAT","LOGLEVEL", \
 "USESYSLOG","SYSLOGNAME","SYSLOGFACILITY","SYSLOGLEVEL","SYSLOGVIRUSLEVEL","IGNOREVIRUS", \
 "DISPLAYINITIALMESSAGES","DBRELOAD","SCANTEMPFILE","TEMPLATEPATH","DISABLELOCKINGFOR", \
 "PARENTPROXY","PARENTPORT","MAXSERVERS","FORWARDED_IP","X_FORWARDED_FOR","FAILSCANERROR", \
 "SSLTIMEOUT", \
 "MAXDOWNLOADSIZE","SCANNERTIMEOUT","STREAMUSERAGENT","STREAMSCANSIZE","SCANIMAGES", \
 "SKIPMIME","SCANMIME", \
 "ENABLECLAMLIB","CLAMDBDIR","CLAMBLOCKBROKEN","CLAMBLOCKMAX","CLAMBLOCKENCRYPTED", \
 "CLAMMAXFILES","CLAMMAXFILESIZE","CLAMMAXRECURSION","CLAMMAXSCANSIZE", \
 "ENABLEAVG","AVGSERVER","AVGPORT", \
 "ENABLEAVESERVER","AVESOCKET", \
 "ENABLEFPROT","FPROTSERVER","FPROTPORT","FPROTOPTIONS", \
 "ENABLETROPHIE","TROPHIEMAXFILES","TROPHIEMAXFILESIZE","TROPHIEMAXRATIO", \
 "ENABLENOD32","NOD32SOCKET","NOD32VERSION", \
 "ENABLECLAMD","CLAMDSOCKET","CLAMDSERVER","CLAMDPORT", \
 "ENABLESOPHIE","SOPHIESOCKET", \
 "ENABLEAVAST","AVASTSOCKET","AVASTSERVER","AVASTPORT", \
 "ENABLEARCAVIR","ARCAVIRSOCKET","ARCAVIRVERSION", \
 "ENABLEDRWEB","DRWEBSOCKET","DRWEBSERVER","DRWEBPORT","DRWEBHEURISTIC","DRWEBMALWARE", \
 "PARENTUSER", "PARENTPASSWORD"
//SCANNERS


//##############################################################
//Configuration not setable in havp.config

//CONNTIMEOUT in seconds
#define CONNTIMEOUT 60

//RECVTIMEOUT in seconds
#define RECVTIMEOUT 120

//SENDTIMEOUT in seconds
#define SENDTIMEOUT 120

//Maximum client connection waiting for accept
#define MAXCONNECTIONS 1024

//Maximum bytes received in one request
#define MAXRECV 14600

//Maximum logfile line length
#define STRINGLENGTH 1000

//Maximum hardlock size - do not change
#define MAXFILELOCKSIZE 1000000000

//Valid Methods
#define METHODS \
 "GET","POST","HEAD","CONNECT","PUT","TRACE","PURGE","OPTIONS","UNLOCK", \
 "SEARCH","PROPFIND","BPROPFIND","PROPPATCH","BPROPPATCH","MKCOL","COPY", \
 "BCOPY","MOVE","LOCK","BMOVE","DELETE","BDELETE","SUBSCRIBE","UNSUBSCRIBE", \
 "POLL","REPORT","ERROR","NONE","MKACTIVITY","CHECKOUT","MERGE"

//Maximum length of SCANTEMPFILE
#define MAXSCANTEMPFILELENGTH 200

//Maximum length of http headers
#define MAXHTTPHEADERLENGTH 65536

// HTML Error String
#define ERROR_DNS	"dns.html"
#define VIRUS_FOUND	"virus.html"
#define ERROR_SCANNER	"scanner.html"
#define ERROR_DOWN	"down.html"
#define ERROR_INVALID	"invalid.html"
#define ERROR_REQUEST	"request.html"
#define ERROR_BODY	"error.html"
#define ERROR_BLACKLIST	"blacklist.html"
#define ERROR_MAXSIZE	"maxsize.html"

#include "_default.h"

#endif
